tests/common.c

```diff
@@ -0,0 +1,307 @@
+
+  n = send(conn->fd, buf, len, 0);
+  if(n < 0) {
+    return errno;
+  }
+  if(out_n != NULL) {
+    *out_n = n;
+  }
+  return 0;
+}
+
+size_t
+bytevec_available(struct bytevec *vec)
+{
+  return vec->capacity - vec->len;
+}
+
+char *
+bytevec_writeable(struct bytevec *vec)
+{
+  return vec->data + vec->len;
+}
+
+void
+bytevec_consume(struct bytevec *vec, size_t n)
+{
+  vec->len += n;
+}
+
+// Ensure there are at least n bytes available between vec->len and
+// vec->capacity. If this requires reallocating, this may return
+// CRUSTLS_DEMO_ERROR.
+enum crustls_demo_result
+bytevec_ensure_available(struct bytevec *vec, size_t n)
+{
+  size_t available = vec->capacity - vec->len;
+  size_t newsize;
+  void *newdata;
+  if(available < n) {
+    newsize = vec->len + n;
+    if(newsize < vec->capacity * 2) {
+      newsize = vec->capacity * 2;
+    }
+    newdata = realloc(vec->data, newsize);
+    if(newdata == NULL) {
+      fprintf(stderr, "out of memory trying to get %ld bytes\n", newsize);
+      return CRUSTLS_DEMO_ERROR;
+    }
+    vec->data = (char *)newdata;
+    vec->capacity = newsize;
+  }
+  return CRUSTLS_DEMO_OK;
+}
+
+int
+copy_plaintext_to_buffer(struct conndata_t *conn)
+{
+  int result;
+  size_t n;
+  struct rustls_connection *rconn = conn->rconn;
+
+  if(bytevec_ensure_available(&conn->data, 1024) != CRUSTLS_DEMO_OK) {
+    return CRUSTLS_DEMO_ERROR;
+  }
+
+  for(;;) {
+    char *buf = bytevec_writeable(&conn->data);
+    size_t avail = bytevec_available(&conn->data);
+    result = rustls_connection_read(rconn, (uint8_t *)buf, avail, &n);
+    if(result == RUSTLS_RESULT_ALERT_CLOSE_NOTIFY) {
+      fprintf(stderr, "Received close_notify, cleanly ending connection\n");
+      return CRUSTLS_DEMO_CLOSE_NOTIFY;
+    }
+    if(result != RUSTLS_RESULT_OK) {
+      fprintf(stderr, "Error in rustls_connection_read: %d\n", result);
+      return CRUSTLS_DEMO_ERROR;
+    }
+    if(n == 0) {
+      /* This is expected. It just means "no more bytes for now." */
+      return CRUSTLS_DEMO_OK;
+    }
+    bytevec_consume(&conn->data, n);
+  }
+
+  return CRUSTLS_DEMO_ERROR;
+}
+
+/**
+ * Since memmem is not cross-platform compatible, we bring our own.
+ * Copied from https://www.capitalware.com/rl_blog/?p=5847.
+ *
+ * Function Name
+ *  memmem
+ *
+ * Description
+ *  Like strstr(), but for non-text buffers that are not NULL delimited.
+ *
+ *  public domain by Bob Stout
+ *
+ * Input parameters
+ *  haystack    - pointer to the buffer to be searched
+ *  haystacklen - length of the haystack buffer
+ *  needle      - pointer to a buffer that will be searched for
+ *  needlelen   - length of the needle buffer
+ *
+ * Return Value
+ *  pointer to the memory address of the match or NULL.
+ */
+void *
+memmem(const void *haystack, size_t haystacklen, const void *needle,
+       size_t needlelen)
+{
+  const char *bf = (const char *)haystack;
+  const char *pt = (const char *)needle;
```

Comment:
nit
```suggestion
  const char *bf = haystack;
  const char *pt = needle;
```
Same as elsewhere: the casts are only needed if compiling as C++.